src/wkt/src/internal/int32.rs

```diff
@@ -0,0 +1,160 @@
+
+impl<'de> serde_with::DeserializeAs<'de, i32> for I32 {
+    fn deserialize_as<D>(deserializer: D) -> Result<i32, D::Error>
+    where
+        D: serde::de::Deserializer<'de>,
+    {
+        deserializer.deserialize_any(I32Visitor)
+    }
+}
+
+const ERRMSG: &str = "a 32-bit signed integer";
+
+struct I32Visitor;
+
+impl serde::de::Visitor<'_> for I32Visitor {
+    type Value = i32;
+
+    fn visit_str<E>(self, value: &str) -> std::result::Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        // ProtoJSON says that both strings and numbers are accepted. Parse the
+        // string as a f64 number (all JSON numbers are f64) and then try to
+        // parse that as a
+        let number = value.parse::<f64>().map_err(E::custom)?;
+        self.visit_f64(number)
+    }
+
+    fn visit_i64<E>(self, value: i64) -> std::result::Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        match value {
+            _ if value < i32::MIN as i64 => Err(self::value_error(value)),
+            _ if value > i32::MAX as i64 => Err(self::value_error(value)),
+            _ => Ok(value as i32),
+        }
+    }
+
+    fn visit_u64<E>(self, value: u64) -> std::result::Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        match value {
+            _ if value > i32::MAX as u64 => Err(self::value_error(value)),
+            _ => Ok(value as i32),
+        }
+    }
+
+    fn visit_f64<E>(self, value: f64) -> std::result::Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        match value {
+            _ if value < i32::MIN as f64 => Err(self::value_error(value)),
+            _ if value > i32::MAX as f64 => Err(self::value_error(value)),
+            _ if value.fract().abs() > 0.0 => Err(self::value_error(value)),
+            // The number is "rounded towards zero". Because we are in range,
+            // and the fractional part is 0, this conversion should be safe.
+            // See https://doc.rust-lang.org/reference/expressions/operator-expr.html#r-expr.as.numeric.float-as-int
+            _ => Ok(value as i32),
+        }
+    }
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("a 32-bit integer in ProtoJSON format")
+    }
+}
+
+fn value_error<T, E>(value: T) -> E
+where
+    T: std::fmt::Display,
+    E: serde::de::Error,
+{
+    E::invalid_value(Other(&format!("{value}")), &ERRMSG)
+}
+
+impl serde_with::SerializeAs<i32> for I32 {
+    fn serialize_as<S>(source: &i32, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_i32(*source)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use anyhow::Result;
+    use serde_json::{Value, json};
+    use serde_with::{DeserializeAs, SerializeAs};
+    use test_case::test_case;
+
+    #[test_case(0, 0)]
+    #[test_case("0", 0; "zero string")]
+    #[test_case("2.0", 2)]
```

Comment:
Is culture relevant? i.e. could this sometimes be "2,0"?

Reply:
I do not see how localization could possibly work for ProtoJSON, since (1) the sender and receiver may have different locales, and (2) there is nothing in the format to specify the source locale. I can only conclude that it always uses the "C" locale.  The same must be true of plain JSON.

Your question raises a followup: does Rust use the current locale to parse floating point numbers? The documentation is silent, which I read as "no localization", and the Interwebs recommend using external crates for localization (including, gulp!, `icu`).

I think we are safe.

Reply:
PS: I also grew up writing "2,0", and MM-DD-YY still irks me.